### ISSUE TITLE
Local cell tower database location backend

### ISSUE THREAD

**Issue by 33KK** [microg/GmsCore#2067]
**Is your feature request related to a problem? Please describe.**
Current solutions are not very useful in any new areas or when the current local backend remembers cell tower positions incorrectly.

**Describe the solution you'd like**
Basically what this project does: https://gitlab.com/deveee/Local-GSM-Backend
It just downloads cell tower databases from OpenCellID and Mozilla and merges them into a single database.

**Describe alternatives you've considered**
There's no alternative available at the moment

**Additional context**
In terms of licensing, Mozilla dataset is public domain, and OpenCellID is CC-BY-SA-4.0, though it requires an API key. Local GSM Backend also included [lacells](https://github.com/sobrus/FastLacellsGenerator/blob/master/flg) as a source, which is just a script that merges data from those two sources, plus radiocells.org (which seems to be dead now). Reuploading the database somewhere to avoid the API key requirement should be fine. Also lacells allows for downloading country specific datasets.

**Comment by CoelacanthusHex** [comment]
microG has used Mozilla Location Service as the source, so it may be enough to add OpenCellID as another source.

**Comment by Iey4iej3** [comment]
Basically https://github.com/microg/GmsCore/issues/1941

**Comment by Iey4iej3** [comment]
> It's what the "Remember from GPS" checkbox already does.

No, `LocalNLPBackend` also supports importing database. It does not implement downloading database, but I use https://github.com/sobrus/FastLacellsGenerator to do this (which is more reliable, and this is suggested by the author of Local-GSM-Backend: https://github.com/n76/lacells-creator).

**Comment by IzzySoft** [comment]
@Iey4iej3 you say `LocalNLPBackend` can be used the same way one used [Local-GSM-Backend](https://gitlab.com/deveee/Local-GSM-Backend) in the past – i.e. one can simply push the SQLite database to the device and it's being used automatically (and that can be done already now with the current release of microG)? Or does it need "manual steps" to import the database?

**Comment by Iey4iej3** [comment]
> @Iey4iej3 you say `LocalNLPBackend` can be used the same way one used [Local-GSM-Backend](https://gitlab.com/deveee/Local-GSM-Backend) in the past – i.e. one can simply push the SQLite database to the device and it's being used automatically (and that can be done already now with the current release of microG)? Or does it need "manual steps" to import the database?

1. LocalNLPBackend has to import the database.
2. I am not sure whether I am mistaken: it does not seem to be able to import SQL database, but plain text `csv` database.
3. I do not make any implication on the current release of microG. I just mention that, the proposal of this functionality was included in https://github.com/microg/GmsCore/issues/1941

**Comment by IzzySoft** [comment]
Thanks for clarification! So you referred to the (temporary) CSV generated by FastLacellsGenerator (which one can configure to keep instead of remove at the end of the process), and this still requires manual steps to import after pushing to the device.

In that case I'll have to be patient and stick with the older microG a bit longer (I push the database to multiple devices automatically and the manual steps on some of them would simply be forgotten to perform – and I don't want to go with an online MLS for several reasons (e.g. it's useless in areas without cell coverage, or on devices without data plan)).

**Comment by Iey4iej3** [comment]
I am using the older microG as well, in order to use LocalNLPBackend (which is, as far as I understood, not completely integrated into the latest microG). I do not remember why I switched from Local-GSM-Backend to LocalNLPBackend. Maybe I wanted to the ability to collect local data in the same time (although I use `Tower Collector` to collect data and upload, it does not seem to improve the precision as well as local collections).

**Comment by IzzySoft** [comment]
TowerCollector is one of my most used apps as well :smile: I've requested once that collections could be shared directly with apps like Local-GSM-Backend, but I don't remember where that got stuck. Probably the broadcast was created with TowerCollector, but nobody wanted to pick it up…

But we're starting to drift towards OT for this issue. Maybe the SQLite-DB created by FastLacellsGenerator could be used directly by microG. That would still mean to create it "elsewhere" of course, but if the current module of Local-GSM-Backend could be used stand-alone, even that could happen on-device – without the problems the "old structure" caused.

**Comment by vchigrin** [comment]
I think it will be very useful to have at least some script for converting OpenCellId.csv file to GmsCore readable `cell-*.csv.gz` file. Seems, at present format of that GmsCore .csv is not documented at it is not trivial to convert one to another.

**Comment by vchigrin** [comment]
Tried something like that https://gist.github.com/vchigrin/97ae1bb9317fa51a13a4d07ae2cc9bf3 but unfortunately after importing result into GmsCore it stopped provide location at all. May be I'll debug this when I have enough time.  At present may be it will be a good starting point for somebody interested.

**Comment by stuart-nolan** [comment]
> For everyone: please post links to specs and samples files of all formats that are needed.

Opencellid: https://wiki.opencellid.org/wiki/Database_format

Thankyou

**Comment by marek22k** [comment]
Is there a OpenCellID to GMS csv converter in the meantime?
